lib/spack/spack/package.py

```diff
@@ -1420,6 +1524,43 @@ def rpath_args(self):
         return " ".join("-Wl,-rpath,%s" % p for p in self.rpath)
 
 
+class Package(PackageBase):
+    phases = ['install', 'log']
+    # This will be used as a registration decorator in user
+    # packages, if need be
+    PackageBase.sanity_check('install')(PackageBase.sanity_check_prefix)
+
+
+class AutotoolsPackage(PackageBase):
+    phases = ['autoreconf', 'configure', 'build', 'install', 'log']
+
+    def autoreconf(self, spec, prefix):
+        """Not needed usually, configure should be already there"""
+        pass
+
+    @PackageBase.sanity_check('autoreconf')
+    def is_configure_or_die(self):
+        if not os.path.exists('configure'):
+            raise RuntimeError('configure script not found in {0}'.format(os.getcwd()))
+
+    def configure_args(self):
+        return list()
+
+    def configure(self, spec, prefix):
+        options = ['--prefix={0}'.format(prefix)] + self.configure_args()
+        inspect.getmodule(self).configure(*options)
+
+    def build(self, spec, prefix):
+        inspect.getmodule(self).make()
+
+    def install(self, spec, prefix):
+        inspect.getmodule(self).make('install')
+
+    # This will be used as a registration decorator in user
+    # packages, if need be
+    PackageBase.sanity_check('install')(PackageBase.sanity_check_prefix)
```

Comment:
I still do not understand why this is useful, or how the user would control
things if phases could be named anything and differ between packages.
On Jul 8, 2016 9:58 AM, "Massimiliano Culpo" <notifications@github.com>
wrote:

> In lib/spack/spack/package.py:
>
> > +    def configure_args(self):
> > +        return list()
> > +
> > +    def configure(self, spec, prefix):
> > +        options = ['--prefix={0}'.format(prefix)] + self.configure_args()
> > +        inspect.getmodule(self).configure(*options)
> > +
> > +    def build(self, spec, prefix):
> > +        inspect.getmodule(self).make()
> > +
> > +    def install(self, spec, prefix):
> > +        inspect.getmodule(self).make('install')
> > +
> > +    # This will be used as a registration decorator in user
> > +    # packages, if need be
> > +    PackageBase.sanity_check('install')(PackageBase.sanity_check_prefix)
>
> @tgamblin <https://github.com/tgamblin> @citibeth
> <https://github.com/citibeth> @adamjstewart
> <https://github.com/adamjstewart> @davydden <https://github.com/davydden>
> @nrichart <https://github.com/nrichart> Here's a first stub at how a
> specialized class may look like right now :
>
>    - phases are described with a list of strings.
>    - default implementations may (but need not to) be provided for every
>    phase
>    - methods can be registered at package definition time as either
>    precondition or sanity_check for any number of phases
>
> Comments ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/1186/files/fa2783b9dc8c13e675a9b6791b240cdca61e4307#r70077657>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AB1cdy593mJztaO0eUhCWChaymSIb9Ihks5qTld_gaJpZM4JG8iX>
> .
>